### PR TITLE
Sprint 37

### PR DESCRIPTION
**Pull request by ericsorenson**
# Sprint 37

### [3926 Basic Search for Orders (system-created) (In progress)](https://github.com/flexion/ef-cms/issues/3926/)

Status checks need to be updated. build-client is being replaced by `build-client-coverage`, `build-client-integration`, and `build-client-unit`.
image

### CODE REVIEW

**Comment by kfoley-18F** [comment]
We use this list when performing a code review to ensure that all tasks have been completed.

- [ ] begin by deleting any existing copy of the Git repository, running `docker system prune` to remove any cached artifacts from any prior code review
- [ ] clone a fresh copy of the Git repository (`git clone https://github.com/ustaxcourt/ef-cms.git`), fetch the pull request and name it for the sprint (e.g., `git fetch origin pull/204/head:sprint-43`), and then switch to that branch (e.g. `git checkout sprint-43`)
- [ ] review the pull request itself, to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] stand up the site locally, with `./docker-run.sh`
	- [ ] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [ ] for internal Court functionality, perform the most thorough testing in Chrome, though also test in Edge and Firefox
		- [ ] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/), including Internet Explorer 11
		- [ ] test in Mobile Safari and Mobile Chrome, with the caveat that not all internal Court functionality will be necessary on these platforms
	- [ ] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [ ] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [ ] review for accessibility
		- [ ] use an automated audit tool, such as Chrome Audit or aXe
		- [ ] navigate site only with the keyboard
		- [ ] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [ ] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [ ] [review static code analysis results](https://sonarcloud.io/organizations/flexion-github/projects) in the vendor’s SonarCloud environment
- [ ] examine OWASP ZAP output in `docs/`, to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, [looking for problems](#what-we-look-for) (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] look at code coverage of tests in [the vendor's SonarCloud environment](https://sonarcloud.io/organizations/flexion-github/projects)
	- [ ] methodically review all new tests for correctness, quality of naming
- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

File: proxy-request-times.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "undefined": {
+    "requests": 5482,
```
[the remark this answers is not in the document]
Comment (sutt0n) [reply proxy-request-times.json:3]: This file should not have been committed, I believe. It's in the `.gitignore` file, but it's being missed by the filter for some reason.

File: README.md [hunk 2]
```diff
@@ -128,6 +136,7 @@ privatePractitioner
 privatePractitioner1 - privatePractitioner4
 irsPractitioner
 irsPractitioner1 - irsPractitioner4
+irsSuperuser
```
[the remark this answers is not in the document]
Comment (sutt0n) [reply README.md:139]: Roles and permissions aren't specifically documented, but the authorization map of them are in `shared/src/authorization/authorizationClientService.js`. Would you like for us to document these in a table or chart of some kind?

The verbiage of the role of `irsSuperuser` is what was agreed upon by the Court.

File: shared/src/authorization/authorizationClientService.js [hunk 3]
```diff
@@ -117,6 +117,12 @@ const AUTHORIZATION_MAP = {
     ROLE_PERMISSIONS.UPLOAD_DOCUMENT,
     ROLE_PERMISSIONS.VIEW_DOCUMENTS,
   ],
+  irsSuperuser: [
+    ROLE_PERMISSIONS.ADVANCED_SEARCH,
+    ROLE_PERMISSIONS.GET_CASE,
+    ROLE_PERMISSIONS.VIEW_DOCUMENTS,
+    ROLE_PERMISSIONS.VIEW_SEALED_CASE,
+  ],
```
[the remark this answers is not in the document]
Comment (sutt0n) [reply shared/src/authorization/authorizationClientService.js:125]: Each permission used in an interactor/use-case _should_ have an associated unit test with its associated authorization check, and this is reflected by SonarCloud's report for test coverage on the specific authorization mapping. Each lambda function calls an interactor, and _most_ (except those on the public API stack) of these interactors called by a lambda have an authorization check associated with it. Each permission can be checked by searching for its associated `ROLE_PERMISSIONS.<PERMISSION_NAME>` in `shared/src/business/useCases`.

The tests are spread out, because each permission check is directly associated with the action it's checking prior to executing it. If you have a specific example, we can walk you through the process all the way from the Serverless config of a given stack down to the permission being checked.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfoley-18F) [reply README.md:139]: Some documentation about roles and permissions would be helpful, thanks!

File: shared/src/authorization/authorizationClientService.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kfoley-18F) [reply shared/src/authorization/authorizationClientService.js:125]: This is great explanation, thanks!

**Comment by kfoley-18F** [comment]
@sutt0n @ericsorenson - looks like the CircleCI install failed on the last run on a plugin install in terraform, due to network error. 

Can you please kick a new build?

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfoley-18F) [reply README.md:139]: (But can be a backlog item)

**Comment by kfoley-18F** [comment]
@JessicaMarine - I recommend this PR be merged into the Court's repo.
